search/includes/classes/class-settingshealthjob.php

```diff
@@ -200,4 +220,84 @@ public function send_alert( $channel, $message, $level, $type = '' ) {
 
 		return wpcom_vip_irc( $channel, $message, $level );
 	}
+
+	/**
+	 * Determine whether to build new index as part of auto healing.
+	 *
+	 * @param object $indexable The Indexable we want to rebuild.
+	 */
+	public function maybe_build_new_index( $indexable ) {
+		// Only do for non-production for now.
+		$is_prod = defined( 'VIP_GO_APP_ENVIRONMENT' ) && 'production' === VIP_GO_APP_ENVIRONMENT;
+		if ( $is_prod ) {
+			return;
+		}
+
+		// Check if lock is set. Bail if new build is already ongoing.
+		$new_index_lock = get_option( 'vip_search_new_version_building_for_shards' );
+		if ( false === $new_index_lock ) {
+			// No lock, check if event has been scheduled already.
+			if ( ! wp_next_scheduled( self::CRON_EVENT_BUILD_NAME ) ) {
+				$scheduled_event = wp_schedule_single_event( time() + 30, self::CRON_EVENT_BUILD_NAME, [ $indexable ] );
+
+				if ( is_wp_error( $scheduled_event ) ) {
+					// Send alert that failed to schedule event.
+					$error_message = $scheduled_event->get_error_message() ?? 'Event could not be scheduled';
+					$message       = sprintf( 'Failed to schedule event for building new %s index version on %s to meet shard requirements: %s', $indexable->slug, home_url(), $error_message );
+					$this->send_alert( '#vip-go-es-alerts', $message, 2 );
+				}
+			}
+		}
+	}
+
+	/**
+	 * Build new index and hot-swap it afterwards as part of auto healing to ensure shard requirements are met.
+	 *
+	 * @param object $indexable The Indexable we want to rebuild.
+	 */
+	public function build_new_index( $indexable ) {
+		$search           = \Automattic\VIP\Search\Search::instance();
+		$current_versions = $search->versioning->get_versions( $indexable );
+
+		// Do not attempt build if limit is reached (2 versions per Indexable).
+		if ( count( $current_versions ) > 1 ) {
+			$message = sprintf(
+				'Cannot automatically build new %s index on %s to meet shard requirements. Please ensure there is less than 2 %s index versions.',
+				$indexable->slug,
+				home_url(),
+				$indexable->slug,
+			);
+			$this->send_alert( '#vip-go-es-alerts', $message, 2 );
+			
+			return;
+		}
+
+		update_option( self::BUILD_LOCK_NAME, time() ); // Set lock for starting rebuild.
+
+		// Do the indexing.
+		$assoc_args = [
+			'using-versions' => true,
+			'skip-confirm'   => true,
+			'indexables'     => $indexable->slug,
+		];
+		$cmd        = 'vip-search index ' . WP_CLI\Utils\assoc_args_to_str( $assoc_args );
+		$result     = WP_CLI::runcommand(
```

Comment:
Is it possible to call the underlining functions rather than running the CLI in a child process?

Reply:
@WPprodigy That was my initial approach but then, I realized that I wanted to capture any errors returned. To accomplish that  I think the underlining function would require re-factoring to capture any errors to be used outside a non-CLI context.

Reply:
Gotcha. Should be fine, just typically worth avoiding the extra complexity if possible.